### PR TITLE
feat: add merge_insert to the node and rust APIs

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

**Comment by westonpace** [comment]
The unit tests will fail until lance can be upgraded to 0.9.12 due to https://github.com/lancedb/lance/pull/1899

File: rust/vectordb/src/table.rs [hunk 1]
```diff
@@ -170,6 +176,71 @@ pub trait Table: std::fmt::Display + Send + Sync {
...
+    ///    ].into_iter().map(Ok),
+    ///   schema.clone());
+    /// // Perform an upsert operation
+    /// let mut merge_insert = tbl.merge_insert(&["id"]);
+    /// merge_insert.when_matched_update_all()
+    ///             .when_not_matched_insert_all();
+    /// merge_insert.execute(Box::new(new_data)).await.unwrap();
+    /// # });
+    /// ```
+    fn merge_insert(&self, on: &[&str]) -> MergeInsertBuilder;
```
[the remark this answers is not in the document]
Comment (westonpace) [reply rust/vectordb/src/table.rs:242]: Upsert has well defined semantics and is only a subset of merge insert.  Merge insert can do `upsert`, `insert-if-not-exists` and `replace-range`.  We could create an alias that sets the appropriate configuration and runs merge insert but we talked about it a bit in slack and decided not to.

[the remark this answers is not in the document]
Comment (westonpace) [reply rust/vectordb/src/table.rs:242]: More specifically:

### Bulk Update

```
when_matched_update_all
```

### Upsert

```
when_matched_update_all
when_not_matched_insert_all
```

### Insert if not Exists

```
when_not_matched_insert_all
```

### Replace Range

```
when_matched_update_all
when_not_matched_insert_all
when_not_matched_by_source_delete
```

[the remark this answers is not in the document]
Comment (wjones127) [reply rust/vectordb/src/table.rs:242]: And, to be clear, we didn't invent `MERGE INSERT`. It's an established operation in data warehouses:

* [Redshift](https://docs.aws.amazon.com/redshift/latest/dg/r_MERGE.html)
* [Databricks](https://docs.databricks.com/en/sql/language-manual/delta-merge-into.html)
* [Snowflake](https://docs.snowflake.com/en/sql-reference/sql/merge)

**Review by albertlockett — APPROVED** [review]

**Pull request merged**